// appium-xcode/build/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.getMaxTVOSSDKWithoutRetry = exports.getMaxTVOSSDK = exports.getCommandLineToolsVersion = exports.getInstrumentsPath = exports.clearInternalCache = exports.getConnectedDevices = exports.getMaxIOSSDKWithoutRetry = exports.getAutomationTraceTemplatePathWithoutRetry = exports.getMaxIOSSDK = exports.getAutomationTraceTemplatePath = exports.getVersion = exports.getPath = void 0;

var xcode = _interopRequireWildcard(require("./lib/xcode"));

const getPath = xcode.getPath,
      getVersion = xcode.getVersion,
      getAutomationTraceTemplatePath = xcode.getAutomationTraceTemplatePath,
      getMaxIOSSDK = xcode.getMaxIOSSDK,
      getAutomationTraceTemplatePathWithoutRetry = xcode.getAutomationTraceTemplatePathWithoutRetry,
      getMaxIOSSDKWithoutRetry = xcode.getMaxIOSSDKWithoutRetry,
      getConnectedDevices = xcode.getConnectedDevices,
      clearInternalCache = xcode.clearInternalCache,
      getInstrumentsPath = xcode.getInstrumentsPath,
      getCommandLineToolsVersion = xcode.getCommandLineToolsVersion,
      getMaxTVOSSDK = xcode.getMaxTVOSSDK,
      getMaxTVOSSDKWithoutRetry = xcode.getMaxTVOSSDKWithoutRetry;
exports.getMaxTVOSSDKWithoutRetry = getMaxTVOSSDKWithoutRetry;
exports.getMaxTVOSSDK = getMaxTVOSSDK;
exports.getCommandLineToolsVersion = getCommandLineToolsVersion;
exports.getInstrumentsPath = getInstrumentsPath;
exports.clearInternalCache = clearInternalCache;
exports.getConnectedDevices = getConnectedDevices;
exports.getMaxIOSSDKWithoutRetry = getMaxIOSSDKWithoutRetry;
exports.getAutomationTraceTemplatePathWithoutRetry = getAutomationTraceTemplatePathWithoutRetry;
exports.getMaxIOSSDK = getMaxIOSSDK;
exports.getAutomationTraceTemplatePath = getAutomationTraceTemplatePath;
exports.getVersion = getVersion;
exports.getPath = getPath;
var _default = xcode;
exports.default = _default;require('source-map-support').install();
